File: mobile/src/main/java/com/zacharytamas/often/models/Habit.java
package com.zacharytamas.often.models;

import com.orm.SugarRecord;
import com.zacharytamas.often.utils.Dates;

import org.joda.time.DateTime;

import java.util.Date;

/**
 * Created by zacharytamas on 10/25/15.
 */
public class Habit extends SugarRecord<Habit> {
    public String title = "";
    public Boolean required = false;
    public int repeatType = 0;
    public int repeatUnit = 0;
    public int repeatScalar = 0;
    public int repeatWeekdays = 0;
    public Date createdAt = DateTime.now().toDate();
    public Date availableAt;
    public Date lastCompletedAt;
    public Boolean dueAtSpecificTime = false;
    public Date dueAt;
    public int streakValue = 0;

    public Boolean getRepeatsOnWeekday(int day) {
        return Dates.getBitForWeekday(this.repeatWeekdays, day);
    }

    public void setRepeatsOnWeekday(int day, Boolean repeats) {
        this.repeatWeekdays = Dates.setBitForWeekday(this.repeatWeekdays, day, repeats);
    }

    public void completeHabit() {
        Date now = DateTime.now().toDate();
        this.availableAt = Dates.nextAvailableAt(this, now);
        this.lastCompletedAt = now;

        if (required) {
            // If the Habit is due now or after now, this completion will increase the streakValue.
            if (dueAt.after(now) || dueAt.equals(now)) {
                streakValue += 1;
            } else {  // If not, they broke the streak, reset to 0.
                streakValue = 0;
            }
        }
    }
}

File: mobile/src/main/java/com/zacharytamas/often/models/Habit.kt
package com.zacharytamas.often.models

import com.orm.SugarRecord
import com.zacharytamas.often.utils.Dates
import java.util.*

class Habit() : SugarRecord<Habit>() {

    var title: String = ""
    var required: Boolean = true

    var repeatType: Byte = 0
    var repeatUnit: Int = 0
    var repeatScalar: Int = 0
    var repeatWeekdays: Byte = 0

    var createdAt: Date? = Date()
    var availableAt: Date? = null
    var lastCompletedAt: Date? = null
    var dueAtSpecificTime: Boolean = false
    var dueAt: Date? = null
    var streakValue: Int = 0

    fun getRepeatOnWeekday(day: Int): Boolean {
        return Dates.getBitForWeekday(this.repeatWeekdays, day);
    }

    fun setRepeatOnWeekday(day: Int, repeats: Boolean) {
        this.repeatWeekdays = Dates.setBitForWeekday(this.repeatWeekdays, day, repeats);
    }
}
